### PR TITLE
Feature/fast executor

### PR DESCRIPTION
**Pull request by reyoung**

### CODE REVIEW

File: paddle/fluid/framework/details/fast_threaded_ssa_graph_executor.cc [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  }
+
+  PrepareAtomicOpDeps();
+}
+
+FeedFetchList FastThreadedSSAGraphExecutor::Run(
+    const std::vector<std::string> &fetch_tensors) {
+  std::unique_ptr<std::unordered_map<OpHandleBase *, std::atomic<int>>>
+      op_deps = atomic_op_deps_.get();
+  PrepareAtomicOpDeps();
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/fast_threaded_ssa_graph_executor.cc:52]: Why do you call `PrepareAtomicOpDeps` again here?

File: paddle/fluid/framework/details/fast_threaded_ssa_graph_executor.cc [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    RunOpAsync(op_deps.get(), op, &complete_q);
+  }
+
+  while (num_complete != op_deps->size()) {
+    size_t num_comp = complete_q.Pop();
+    if (num_comp == -1UL) {
+      int remaining = remaining_;
+      for (int i = 0; i < remaining; ++i) {
+        complete_q.Pop();
+      }
```
Comment (chengduoZH) [inline paddle/fluid/framework/details/fast_threaded_ssa_graph_executor.cc]: Why do you `Pop` elements of `complete_q` when an exception was encountered? is it necessary?

File: paddle/fluid/framework/details/fast_threaded_ssa_graph_executor.cc (hunk 1 above)
Comment (reyoung) [reply paddle/fluid/framework/details/fast_threaded_ssa_graph_executor.cc:52]: It is prepared for next Run in async.

File: paddle/fluid/framework/details/fast_threaded_ssa_graph_executor.cc (hunk 2 above)
Comment (reyoung) [reply paddle/fluid/framework/details/fast_threaded_ssa_graph_executor.cc]: It will wait for all workers complete.

Comment (chengduoZH) [reply paddle/fluid/framework/details/fast_threaded_ssa_graph_executor.cc]: I have some questions here:
1. after you call `int remaining = remaining_;`, new threads maybe launched, so `remaining` maybe not the number of running threads.
2. It seems that after calling line105~107 doesn't' mean all the workers have completed.

Comment (reyoung) [reply paddle/fluid/framework/details/fast_threaded_ssa_graph_executor.cc]: Done.

**Review by chengduoZH — APPROVED** [review]
Excellent

**Pull request merged**